### PR TITLE
Lite: Slice Op 4-D bottleneck removed

### PR DESCRIPTION
**Pull request by ANSHUMAN87**
1:> Removed hard coding or bottleneck to 4-Dim
2:> Add assumption, that begin size and input dim are equal

### CODE REVIEW

**Comment by ANSHUMAN87** [comment]
@liyunlu0618 : Gentle Reminder!

**Comment by jdduke** [comment]
If we relax this, we should also update optimized_ops.h. And the optimized path should avoid heap allocation.

**Comment by ANSHUMAN87** [comment]
> If we relax this, we should also update optimized_ops.h. And the optimized path should avoid heap allocation.

@jdduke : I have handled your comments, please help review, Thanks!

**Review by jdduke — CHANGES_REQUESTED** [review]

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: @jdduke: I think Slice Op does not support scalar case, please correct me if i am wrong, Thanks!

File: tensorflow/lite/kernels/slice.cc
Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/slice.cc]: @jdduke : Sorry for confusion, actually Slice Kernel has 2 things to handle.
  1:> Support arbitrary dimension Input.
  2:> Compulsory conversion of all type of Dims input to 4-D.
So in my current PR, i have addressed the 2nd issue, where i made the kernels independent of Input Dimensions, now it can operate on any N-Dims Input.

Once this PR gets approved, i will resolve the 1st issue as well.

**Comment by ANSHUMAN87** [comment]
@jdduke: Your comments and queries are addressed, please check, Thanks!

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (jdduke) [inline tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: rename to is_input_1d

**Review by jdduke — CHANGES_REQUESTED** [review]

**Comment by ANSHUMAN87** [comment]
@jdduke : Your comments are addressed now, please check, Thanks!

**Comment by ANSHUMAN87** [comment]
@jdduke : The conflicts are resolved now, Thanks!

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by gbaned** [comment]
@ANSHUMAN87 Could you please resolve the conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by ANSHUMAN87** [comment]
Resolved!

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (jdduke) [inline tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: Can you run the clang formatter tool on this change? Thanks!

https://www.tensorflow.org/community/contribute/code_style#c_coding_style

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (jdduke) [inline tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: Is WriteN the only real optimization here? If so, I'd be more than OK using WriteN in the reference impl, and removing this version.

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (jdduke) [inline tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: Have you measured performance for the 4D case?

File: tensorflow/lite/kernels/slice_test.cc
Comment (jdduke) [inline tensorflow/lite/kernels/slice_test.cc]: We already support 4D? Shouldn't there be something like a 5D test?

**Review by jdduke — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@ANSHUMAN87 Can you please check reviewer comments and keep us posted. Thanks!

File: tensorflow/lite/kernels/slice_test.cc
Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/slice_test.cc]: Please refer: https://github.com/tensorflow/tensorflow/pull/27113#discussion_r271167137

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: @jdduke : The main difference is in Reference(It will iterate through all axis, then fill the output buffer), but in Optimized(It will iterate only axis - 1, and last axis it will just copy the length directly to output buffer).
Please let me know, if still want to merge to one into Reference.
I also feel, it should be moved to reference.

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: I will update you on this!

**Comment by ANSHUMAN87** [comment]
@jdduke : I have replied to your comments inline, please check and feedback, Thanks!

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (jdduke) [reply tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: Any updates?

Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: Please find attached the bench-marking report.
Input shape: [400, 48000, 40]
Output Shape: [400, 35000, 35]
Slice Op input:
slice_output = tf.slice(inputs, (0, 0, 0), (-1, 35000, 35))

Base OP:
SLICE-base-op
Latest Changed OP:
SLICE-new-Change

Both seems same.
Please let me know your opinion, Thanks!

**Comment by ANSHUMAN87** [comment]
@jdduke: I have replied to one of your comment with bench-marking report, please check, Thanks!

**Comment by jdduke** [comment]
Thanks for benchmarking!

I appreciate all the work you've done on this PR, but I'm still struggling to see what improvement this is offering. It doesn't appear any faster, and handles the same 1D-4D cases as the current implementation. If you had a compelling model or use-case which required 5D, I would definitely consider landing this, but until then the benefit isn't as clear. Note that we *do* need a 5D strided_slice implementation, and have encountered a number of models which require that op be extended.

**Comment by ANSHUMAN87** [comment]
> Thanks for benchmarking!
> 
> I appreciate all the work you've done on this PR, but I'm still struggling to see what improvement this is offering. It doesn't appear any faster, and handles the same 1D-4D cases as the current implementation. If you had a compelling model or use-case which required 5D, I would definitely consider landing this, but until then the benefit isn't as clear. Note that we _do_ need a 5D strided_slice implementation, and have encountered a number of models which require that op be extended.

@jdduke : Thanks for your response! I understand your point quite clearly. In fact this PR conceptualized from the need of STRIDED_SLICE(#26748) . Also this changes is not performance improvement, it is only for generalization to support N-dim. The benchmark i shared is to showcase, that there is no performance dip. Another point which i want to bring notice to you. Both this PR and #26748 are just first phase,  i have plan to remove complete dependency in 4-Dim for both SLICE & STRIDED_SLICE in following PRs(Please check my previous comments). But only if these PRs get merged, i can do that. So i urge you to reconsider both current PR & #26748. As the need for STRIDED_SLICE & SLICE for N-dim is quite evident. 

NOTE: This PR does not remove 4-dim dependency completely because of common data-structure limitation, which i will take care in my future PRs(applicable for STRIDED_SLICE#26748 also).

**Comment by jdduke** [comment]
Hey @ANSHUMAN87 , apologies again for the delayed response. We're working on a more comprehensive proposal for updating our ops to support 5D shapes in particular. Those appear to be the most common scenarios that aren't yet supported, and we'd rather address those first than handle the more general case (which may degrade performance). I'll let @thaink chime in on specific thoughts here, but I think we can extend our 4D support to 5D here without sacrificing performance.

**Comment by thaink** [comment]
Hi @ANSHUMAN87,
I don't think your PR introduce significant performance overhead. However, looks like you haven't removed the assumption of dims <= 4. In slice.cc, kMaxDim still being set to 4. 

Besides, the implementation looks more complex than it should. I think the implementation for reference and optimized ops could be the same, the reference ops could also use memcpy with cpy_len. 
My proposal about 5D support is still under review, so I'll update about that when it get submitted.

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (thaink) [inline tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: These logic make the code complicated. I think you can remove them by just removing the for loop in the case axis == final_axis.

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (thaink) [inline tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: In Google coding style, we just use const reference or pointer, reference is not allowed. If you follow the following comment, I think const tflite::SliceParams& will be just fine.

**Comment by ANSHUMAN87** [comment]
> Hi @ANSHUMAN87,
> I don't think your PR introduce significant performance overhead. However, looks like you haven't removed the assumption of dims <= 4. In slice.cc, kMaxDim still being set to 4.
> 
> Besides, the implementation looks more complex than it should. I think the implementation for reference and optimized ops could be the same, the reference ops could also use memcpy with cpy_len.
> My proposal about 5D support is still under review, so I'll update about that when it get submitted.

@thaink : Thanks for your valuable feedback! Just wanted to confirm whether should i wait for your changes or should handle your comments?

**Comment by thaink** [comment]
@ANSHUMAN87, You can handle the comments now since this op is quite exceptional for my changes.

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (thaink) [reply tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: My suggested raw implementation like:
  int final_axis = input_shape.DimensionsCount() - 1;
  int copy_len = op_params.size[final_axis] - op_params.begin[final_axis];

  // Store as std::function to allow recursion.
  std::function<void(int, int)> compute_slice =
      [&compute_slice, final_axis, copy_len, writer,
       input_shape, op_params](int axis, int curr_indice) {
        if (axis == (final_axis)) {
            if (copy_len > 0) writer->WriteN(curr_indice + op_params.begin[final_axis], copy_len);
        } else {
          for (int in_x = op_params.begin[axis]; in_x < op_params.size[axis];
               in_x++) {
            int indices = (in_x + curr_indice) * input_shape.Dims(axis + 1);
            compute_slice(axis + 1, indices);
          }
        }
      };

which can be shared for both reference and optimized_ops. However, I haven't debug this. Could you take time to implement this?

**Comment by ANSHUMAN87** [comment]
> @ANSHUMAN87, You can handle the comments now since this op is quite exceptional for my changes.

@thaink : Thanks for your valuable comments. I have handled all now. Please check. Thanks!

File: tensorflow/lite/kernels/internal/reference/reference_ops.h
Comment (thaink) [inline tensorflow/lite/kernels/internal/reference/reference_ops.h]: I wonder why do we have to avoid WriteN here?

**Comment by thaink** [comment]
With kMaxDim is still 4, generally the PR looks fine to me. However, I tried an approach similar to your lambda implementation recently and found out that it is about 4 times slower than the loop-based one. So let me measure the performance once before merging this PR.

File: tensorflow/lite/kernels/internal/reference/reference_ops.h
Comment (ANSHUMAN87) [reply tensorflow/lite/kernels/internal/reference/reference_ops.h]: @thaink : This was old code, so i did not modify the base logic. However i agree, both optimized & reference should be same. If you agree i can make both optimized and reference same, and remove optimized ops. So that there will be only reference ops code.

**Comment by ANSHUMAN87** [comment]
> With kMaxDim is still 4, generally the PR looks fine to me. However, I tried an approach similar to your lambda implementation recently and found out that it is about 4 times slower than the loop-based one. So let me measure the performance once before merging this PR.

@thaink : I had shared one bench-marking result earlier. Hope it might help you!

**Comment by thaink** [comment]
> > With kMaxDim is still 4, generally the PR looks fine to me. However, I tried an approach similar to your lambda implementation recently and found out that it is about 4 times slower than the loop-based one. So let me measure the performance once before merging this PR.
> 
> @thaink : I had shared one bench-marking result earlier. Hope it might help you!

Thanks. I also get the same result. Can you sync this PR the latest master?

File: tensorflow/lite/kernels/internal/reference/reference_ops.h
Comment (thaink) [reply tensorflow/lite/kernels/internal/reference/reference_ops.h]: Agreed. That is how we usually do.

**Comment by thaink** [comment]
@jdduke Since the kMaxdim is still 4, there should be no effect on delegates. I think it is fine to merge this PR, can you take a look?

File: tensorflow/lite/kernels/internal/reference/reference_ops.h
Comment (jdduke) [reply tensorflow/lite/kernels/internal/reference/reference_ops.h]: Can we make compute_slice a helper function rather than std::function? I think we generally try to avoid using std::function in our reference ops, as they're shared with TFLite Micro.

**Review by jdduke — CHANGES_REQUESTED** [review]
One last request, other lgtm (and thanks again for your patience).

**Comment by gbaned** [comment]
@ANSHUMAN87 Can you please check jdduke's comments and keep us posted. Thanks!

**Comment by ANSHUMAN87** [comment]
@jdduke: Thanks for your comment! 
@gbaned :  I will handle this PR possibly by Saturday. Thanks!

**Comment by ANSHUMAN87** [comment]
@jdduke : Your comment is handled now. Please check. Thanks!

**Review by jdduke — APPROVED** [review]

**Comment by jdduke** [comment]
This is breaking a number of internal tests. At this point, I think it may not be worth the hassle as it's taken nearly a year to get to this point. Thanks for the contribution, we can revisit this PR when it becomes

**Pull request closed without merging**
